man/msgchk.man
.TH MSGCHK %manext1% 2016-11-02 "%nmhversion%"
.
.\" %nmhwarning%
.
.SH NAME
msgchk \- nmh's check for incoming email
.SH SYNOPSIS
.HP 5
.na
.B msgchk
.RB [ \-help ]
.RB [ \-version ]
.RB [ \-date " | " \-nodate ]
.RB [ \-notify
all/mail/nomail ]
.RB [ \-nonotify
all/mail/nomail ]
.RB [ \-host
.IR hostname ]
.RB [ \-user
.IR username ]
.RB [ \-sasl " | " \-nosasl ]
.RB [ \-saslmech
.IR mechanism ]
.RB [ \-tls ]
.RB [ \-initialtls ]
.RB [ \-notls ]
.RB [ \-certverify " | " \-nocertverify ]
.RB [ \-authservice
.IR service ]
.RB [ \-snoop ]
.RI [ users
\&... ]
.ad
.SH DESCRIPTION
The
.B msgchk
program checks all known mail drops for mail waiting
for you.  For those drops which have mail for you,
.B msgchk
will
indicate if it believes that you have seen the mail in question before.
.PP
The
.B \-notify
.I type
switch indicates under what circumstances
.B msgchk
should produce a message.  The default is
.B \-notify
.I all
which says that
.B msgchk
should always report the status of the
users mail drop.  Other values for `type' include `mail' which says that
.B msgchk
should report the status of waiting mail; and, `nomail'
which says that
.B msgchk
should report the status of empty mail drops.
The
.B \-nonotify
.I type
switch has the inverted sense, so
.B \-nonotify
.I all
directs
.B msgchk
to never report the status of
mail drops.  This is useful if the user wishes to check
.BR msgchk 's
exit status.  A non-zero exit status indicates that mail was
.I not
waiting for at least one of the indicated users.
.PP
If
.B msgchk
produces output, then the
.B \-date
switch directs
.B msgchk
to print out the last date mail was read, if this can
be determined.
.SS "Using POP"
.B msgchk
will normally check all the local mail drops, but if
the option \*(lqpophost:\*(rq is set in the mts configuration file
\*(lqmts.conf\*(rq, or if the
.B \-host
.I hostname
switch is given,
.B msgchk
will query this POP service host as to the status of
mail waiting.
.PP
To specify a username for authentication with the POP server, use the
.B \-user
.I username
switch.  The credentials profile entry in
.IR mh\-profile (5)
describes the ways to supply a username and password.
.PP
For debugging purposes, there is also a switch
.BR \-snoop ,
which will
allow you to watch the POP transaction take place between you and the
POP server.  If
.B \-sasl \-saslmech xoauth2
is used, the HTTP transaction is also shown.
.PP
If
.B nmh
has been compiled with SASL support, the
.B \-sasl
switch will enable
the use of SASL authentication.  Depending on the SASL mechanism used, this
may require an additional password prompt from the user (but the
.I netrc
file can be used to store this password, as described in
.IR mh-profile (5).
The
.B \-saslmech
switch can be used to select a particular SASL mechanism.
.PP
If SASL authentication is successful,
.B msgchk
will attempt to negotiate
a security layer for session encryption.  Encrypted traffic is labelled
with `(encrypted)' and `(decrypted)' when viewing the POP transaction
with the
.B \-snoop
switch; see
.IR post (8)'s
description of
.B \-snoop
for its other features.
.PP
If
.B nmh
has been compiled with OAuth support, the
.B \-sasl \-saslmech xoauth2
switch will enable OAuth authentication.  The
.B \-user
switch must be used, and the
.I user-name
must be an email address the user has for the service, which must
be specified with the
.B \-authservice
.I service
switch.  Before using this, the user must authorize nmh by running
.B mhlogin
and grant authorization to that account.  See
.IR mhlogin (1)
for more details.
.PP
If
.B nmh
has been compiled with TLS support, the
.B \-tls
and
.B \-initialtls
switches will require the negotiation of TLS when connecting
to the remote POP server.  The
.B \-tls
switch will negotiate TLS as part of the normal POP protocol using
the STLS command.  The
.B \-initialtls
switch will negotiate TLS immediately after the connection has taken place,
before any POP commands are sent or received.  Data encrypted by TLS is
labeled `(tls-encrypted)' and `(tls-decrypted)' with viewing the POP
transaction with the
.B \-snoop
switch.  The
.B \-notls
switch will disable all attempts to negotiate TLS.
.PP
When using TLS the default is to verify the remote certificate and SubjectName
against the local trusted certificate store.  This can be controlled by
the
.B \-certverify
and
.B \-nocertverify
switches.  See your OpenSSL documentation for more information on certificate
verification.
.SH FILES
.fc ^ ~
.nf
.ta \w'%nmhetcdir%/ExtraBigFileName  'u
^$HOME/.mh_profile~^The user profile
^%nmhetcdir%/mts.conf~^nmh mts configuration file
^%mailspool%/$USER~^Location of mail drop
.fi
.SH "PROFILE COMPONENTS"
.fc ^ ~
.nf
.ta 2.4i
.ta \w'ExtraBigProfileName  'u
None
.fi
.SH "SEE ALSO"
.IR inc (1),
.IR mh\-mail (5),
.IR post (8)
.SH DEFAULTS
.nf
.RB ` user "' defaults to the current user"
.RB ` \-date '
.RB ` "\-notify\ all" '
.fi
.SH CONTEXT
None


NEWS
# NEWS - Release notes for nmh 1.7
#

Welcome to nmh, the new version of the classic MH mail handling system.
It's been over three years since the last release of nmh, and there have
been a number of significant changes since the last release.  Long-time
MH and nmh uses should read careful the NOTEABLE CHANGES section, as there
are some significant changes to nmh behavior.  Otherwise, please see the
README and INSTALL files for help on getting started with nmh.

For news of future releases, subscribe to the low-volume
    https://lists.nongnu.org/mailman/listinfo/nmh-announce

---------------
NOTABLE CHANGES
---------------

The largest notable changes in the 1.7 release are:

- Complete unification of network security support.  All network protocols
  (currently, POP and SMTP) have been refactored to use a common set of
  security routines.  This means all protocols support all SASL mechanisms
  (via the Cyrus-SASL library) and TLS.  TLS support has been strengthened
  to perform certificate name validation and to require TLS 1.1 as a
  minimum protocol.  Also, all protocols can make use of the OAuth2/XOAUTH
  SASL mechanism, which is supported by Gmail.
- send(1) now supports adding switches to post(8) based on the address or
  domain of the email address in the From: header; this more easily allows
  users to support multiple identities.
- A generic facility for passing arguments to filter programs in repl(1)
  by use of the -convertargs switch.
- Native support for the manipulation of iCalendar requests; see mhical(1)
  for more details.

------------
NEW FEATURES
------------

The following are new features for the 1.7 release of nmh:

- When building from source, configure will derive ${prefix} from an existing
  nmh installation if it finds one in your $PATH.
- Added welcome message when nmh detects that its version changed.
- The default locations for configuration files and support binaries
  have been changed.  Configuration files now install into ${sysconfdir}/nmh,
  and support binaries are placed in ${libexecdir}/nmh.  If you are upgrading
  an existing installation you should look for old configuration files in
  ${sysconfdir} and merge any local customizations into the new files in
  ${sysconfdir}/nmh, then remove the old files.  ${libdir} will also contain
  obsolete support programs that should be removed.
- All TLS connections now perform certificate validation (including hostname
  matching) by default; can be disabled on a per-application basis.
- post now defaults to port 587 on 'smtp' message submission.
- A value of 0 for the width switch of scan(1), inc(1), ap(1), dp(1),
  fmttest(1), and mhl(1) now means as many characters as the format
  engine can produce [Bug #15274].  That amount is limited by internal
  buffers.
- If a component has trailing whitespace, e.g., body:component="> ",
  mhl now trims that whitespace off when filtering blank text lines.
- An "rtrim" flag has been added to mhl to remove any trailing
  whitespace from filtered text lines.  A corresponding "nortrim" flag
  has also been added.
- Added getmymbox and getmyaddr mh-format(5) function escapes.
- New -[no]changecur, -fixtype, -decodetypes, and -[no]crlflinebreaks switches
  have been added to mhfixmsg(1).
- mhfixmsg now removes an extraneous trailing semicolon from header
  parameter lists.
- Added -convertargs switch to repl(1), to pass arguments to programs
  specified in the user's profile or mhn.defaults to convert message
  content.
- Added mhical(1), to display, reply to, and cancel iCalendar (RFC 5545)
  event requests.
- Added multiply mh-format(5) function.
- "mhparam bindir" prints the path to the directory containing the public
  executables (${bindir}).
- New "-prefer" switch for mhshow (and mhlist and mhshow), to allow specifying
  the preferred content types to show, if present in a multipart alternative.
- mh-format now has %(kilo) and %(kibi) functions, to allow printing
  numbers with SI or IEC quantities, e.g. "10K", "2.3Mi".
- Support for the -sendmail flag to send/post to change the sendmail
  binary when using the sendmail/pipe MTS.
- Added support to send(1) to specify switches to post(1) based on address or
  domain name in From: header line in message draft.
- post(8) -snoop now attempts to decode base64-encoded SMTP traffic.
- folder(1) -nocreate now prints a warning message for a non-existent folder.
- mhfixmsg(1) now allows -decodetext binary, though 8bit is still the default.
- inc(1) and msgchk(1) now support TLS encryption natively.
- All network protocols support the XOAUTH authentication mechanism.
- Support for SMTPUTF8 (RFC 6531) has been added.  mhshow(1) already supported
  RFC 6532, assuming all 8-bit message header field bodies are UTF-8 and use
  of a UTF-8 locale.
- mhfixmsg now replaces RFC 2047 encoding with RFC 2231 encoding of name and
  filename parameters in Content-Type and Content-Disposition headers,
  respectively.
- If a message body contains 8-bit bytes, post(8) uses SMTP 8BITMIME if the
  server supports it.  If not, post fails with a message to the user to
  encode the message for 7-bit transport.
- Fewer lseek(2)s will be used when reading headers in the common case.
- ./configure's --enable-debug has been removed;  it did nothing.
- configure now defaults to enabling each of TLS and Cyrus SASL if the
  necessary headers and libraries are found.
- Moved build_nmh to top-level directory.
- Better error reporting for connections to network services.

---------
BUG FIXES
---------

- The format scanner no longer subtracts 1 from the width.  This has the
  effect of no longer counting the trailing newline in the output of
  scan(1), inc(1), and the other programs that rely on the format scanner.
- The first character of some very short (less than 4 characters) message
  bodies is no longer dropped.
- Single-character headers can be reliably formatted, etc., instead of
  apparently being missing.

uip/msgchk.c
/* msgchk.c -- check for mail
 *
 * This code is Copyright (c) 2002, by the authors of nmh.  See the
 * COPYRIGHT file in the root directory of the nmh distribution for
 * complete copyright information.
 */

#include "h/mh.h"
#include "sbr/dtime.h"
#include "sbr/getarguments.h"
#include "sbr/smatch.h"
#include "sbr/ambigsw.h"
#include "sbr/print_version.h"
#include "sbr/print_help.h"
#include "sbr/error.h"
#include "h/mts.h"
#include "h/tws.h"
#include "h/done.h"
#include "h/utils.h"
#include <pwd.h>

#include "popsbr.h"

#ifndef CYRUS_SASL
# define SASLminc(a) (a)
#else
# define SASLminc(a)  0
#endif

#ifndef TLS_SUPPORT
# define TLSminc(a) (a)
#else
# define TLSminc(a)  0
#endif

#define MSGCHK_SWITCHES \
    X("date", 0, DATESW) \
    X("nodate", 0, NDATESW) \
    X("notify type", 0, NOTESW) \
    X("nonotify type", 0, NNOTESW) \
    X("host hostname", 0, HOSTSW) \
    X("user username", 0, USERSW) \
    X("port name/number", 0, PORTSW) \
    X("version", 0, VERSIONSW) \
    X("help", 0, HELPSW) \
    X("snoop", 0, SNOOPSW) \
    X("sasl", SASLminc(4), SASLSW) \
    X("nosasl", SASLminc(6), NOSASLSW) \
    X("saslmech", SASLminc(5), SASLMECHSW) \
    X("authservice", SASLminc(0), AUTHSERVICESW) \
    X("tls", TLSminc(-3), TLSSW) \
    X("initialtls", TLSminc(-10), INITTLSSW) \
    X("notls", TLSminc(-5), NOTLSSW) \
    X("certverify", TLSminc(-10), CERTVERSW) \
    X("nocertverify", TLSminc(-12), NOCERTVERSW) \
    X("proxy command", 0, PROXYSW) \

#define X(sw, minchars, id) id,
DEFINE_SWITCH_ENUM(MSGCHK);
#undef X

#define X(sw, minchars, id) { sw, minchars, id },
DEFINE_SWITCH_ARRAY(MSGCHK, switches);
#undef X

/*
 * Maximum numbers of users we can check (plus
 * one for the NULL vector at the end).
 */
#define MAXVEC  51

#define	NT_MAIL	0x1
#define	NT_NMAI	0x2
#define	NT_ALL	(NT_MAIL | NT_NMAI)


/*
 * static prototypes
 */
static int donote (char *, int) PURE;
static int checkmail (char *, char *, int, int, int);
static int remotemail (char *, char *, char *, char *, int, int, int, int,
		       char *, int, const char *);


int
main (int argc, char **argv)
{
    bool datesw = true;
    int notifysw = NT_ALL;
    int status = 0;
    bool sasl = false;
    int tls = 0;
    bool noverify = false;
    bool snoop = false;
    int vecp = 0;
    char *cp, *host = NULL, *port = NULL, *user = NULL, *proxy = NULL;
    char buf[BUFSIZ], *saslmech = NULL, *auth_svc = NULL;
    char **argp, **arguments, *vec[MAXVEC];
    struct passwd *pw;

    if (nmh_init(argv[0], true, true)) { return 1; }

    mts_init ();

    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    while ((cp = *argp++)) {
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
		case AMBIGSW: 
		    ambigsw (cp, switches);
		    done (1);
		case UNKWNSW: 
		    die("-%s unknown", cp);

		case HELPSW: 
		    snprintf (buf, sizeof(buf), "%s [switches] [users ...]",
			invo_name);
		    print_help (buf, switches, 1);
		    done (0);
		case VERSIONSW:
		    print_version(invo_name);
		    done (0);

		case DATESW:
		    datesw = true;
		    continue;
		case NDATESW:
		    datesw = false;
		    continue;

		case NOTESW:
		    if (!(cp = *argp++) || *cp == '-')
			die("missing argument to %s", argp[-2]);
		    notifysw |= donote (cp, 1);
		    continue;
		case NNOTESW:
		    if (!(cp = *argp++) || *cp == '-')
			die("missing argument to %s", argp[-2]);
		    notifysw &= ~donote (cp, 0);
		    continue;

		case HOSTSW: 
		    if (!(host = *argp++) || *host == '-')
			die("missing argument to %s", argp[-2]);
		    continue;

		case PORTSW:
		    if (!(port = *argp++) || *port == '-')
			die("missing argument to %s", argp[-2]);
		continue;

		case USERSW: 
		    if (!(cp = *argp++) || *cp == '-')
			die("missing argument to %s", argp[-2]);
		    if (vecp >= MAXVEC-1)
			die("you can only check %d users at a time", MAXVEC-1);
                    user = vec[vecp++] = cp;
		    continue;

		case SNOOPSW:
		    snoop = true;
		    continue;

		case SASLSW:
		    sasl = true;
		    continue;
		
		case NOSASLSW:
		    sasl = false;
		    continue;

		case SASLMECHSW:
		    if (!(saslmech = *argp++) || *saslmech == '-')
			die("missing argument to %s", argp[-2]);
		    continue;

		case TLSSW:
		    tls = 1;
		    continue;

		case INITTLSSW:
		    tls = 2;
		    continue;

		case NOTLSSW:
		    tls = 0;
		    continue;

		case CERTVERSW:
		    noverify = false;
		    continue;

		case NOCERTVERSW:
		    noverify = true;
		    continue;

		case AUTHSERVICESW:
#ifdef OAUTH_SUPPORT
		    if (!(auth_svc = *argp++) || *auth_svc == '-')
			die("missing argument to %s", argp[-2]);
#else
		    die("not built with OAuth support");
#endif
		    continue;

		case PROXYSW:
		    if (!(proxy = *argp++) || *proxy == '-')
			die("missing argument to %s", argp[-2]);
		    continue;
	    }
	}
	if (vecp >= MAXVEC-1)
	    die("you can only check %d users at a time", MAXVEC-1);
        vec[vecp++] = cp;
    }

    /*
     * If -host is not specified by user
     */
    if (!host || !*host) {
	/*
	 * If "pophost" is specified in mts.conf,
	 * use it as default value.
	 */
	if (pophost && *pophost)
	    host = pophost;
    }
    if (!host || !*host)
	host = NULL;

    if (vecp != 0)
	vec[vecp] = NULL;

    if (host) {
	int tlsflag = 0;

	if (tls == 1)
	    tlsflag = P_STARTTLS;
	else if (tls == 2)
	    tlsflag = P_INITTLS;

	if (noverify)
	    tlsflag |= P_NOVERIFY;

	if (vecp == 0) {
	    status = remotemail (host, port, user, proxy, notifysw, 1,
				 snoop, sasl, saslmech, tlsflag, auth_svc);
	} else {
	    for (vecp = 0; vec[vecp]; vecp++)
		status += remotemail (host, port, vec[vecp], proxy, notifysw, 0,
				      snoop, sasl, saslmech, tlsflag, auth_svc);
	}
    } else {
	if (user == NULL) user = getusername ();
	if (vecp == 0) {
	    char *home;

	    /* Not sure this check makes sense... */
	    if (!geteuid() || NULL == (home = getenv("HOME"))) {
		pw = getpwnam (user);
		if (pw == NULL)
		    die("unable to get information about user");
		home = pw->pw_dir;
	    }
	    status = checkmail (user, home, datesw, notifysw, 1);
	} else {
	    for (vecp = 0; vec[vecp]; vecp++) {
		if ((pw = getpwnam (vec[vecp])))
		    status += checkmail (pw->pw_name, pw->pw_dir, datesw, notifysw, 0);
		else
		    inform("no such user as %s", vec[vecp]);
	    }
	}
    }		/* host == NULL */

    done (status);
    return 1;
}


#define NOTE_SWITCHES \
    X("all", 0, NALLSW) \
    X("mail", 0, NMAISW) \
    X("nomail", 0, NNMAISW) \

#define X(sw, minchars, id) id,
DEFINE_SWITCH_ENUM(NOTE);
#undef X

#define X(sw, minchars, id) { sw, minchars, id },
DEFINE_SWITCH_ARRAY(NOTE, ntswitches);
#undef X


static int
donote (char *cp, int ntflag)
{
    switch (smatch (cp, ntswitches)) {
	case AMBIGSW: 
	    ambigsw (cp, ntswitches);
	    done (1);
	case UNKWNSW: 
	    die("-%snotify %s unknown", ntflag ? "" : "no", cp);

	case NALLSW: 
	    return NT_ALL;
	case NMAISW: 
	    return NT_MAIL;
	case NNMAISW: 
	    return NT_NMAI;
    }

    return 0; /* Before 1999-07-15, garbage was returned if control got here. */
}


static int
checkmail (char *user, char *home, int datesw, int notifysw, int personal)
{
    char buffer[BUFSIZ];
    snprintf(buffer, sizeof buffer, "%s/%s", *mmdfldir ? mmdfldir : home,
        *mmdflfil ? mmdflfil : user);

    struct stat st;
    bool statok = stat(buffer, &st) != -1;
    bool empty = !statok || st.st_size == 0;

    if ((empty && !(notifysw & NT_NMAI)) ||
        (!empty && !(notifysw & NT_MAIL))) {
        return empty;
    }

    if (empty) {
        if (personal) {
            fputs("You don't have any mail waiting", stdout);
        } else {
            printf("%s doesn't have any mail waiting", user);
        }
    } else {
        char *kind = st.st_mtime < st.st_atime ? "old" : "new";
        if (personal) {
            printf("You have %s mail waiting", kind);
        } else {
            printf("%s has %s mail waiting", user, kind);
        }
    }

    if (datesw && statok && st.st_atime)
        printf("; last read on %s", dtime(&st.st_atime, 1));
    putchar('\n');

    return empty;
}


extern char response[];

static int
remotemail (char *host, char *port, char *user, char *proxy, int notifysw,
	    int personal, int snoop, int sasl, char *saslmech, int tls,
	    const char *auth_svc)
{
    int nmsgs, nbytes;
    bool status;

    if (auth_svc == NULL) {
	if (saslmech  &&  ! strcasecmp(saslmech, "xoauth2")) {
	    die("must specify -authservice with -saslmech xoauth2");
	}
    } else {
	if (user == NULL) {
	    die("must specify -user with -saslmech xoauth2");
	}
    }

    /* open the POP connection */
    if (pop_init (host, port, user, proxy, snoop, sasl, saslmech, tls,
		  auth_svc) == NOTOK
	    || pop_stat (&nmsgs, &nbytes) == NOTOK     /* check for messages  */
	    || pop_quit () == NOTOK) {                 /* quit POP connection */
	inform("%s", response);
	return 1;
    }

    if (nmsgs) {
	if (notifysw & NT_MAIL) {
	    if (personal)
		fputs("You have ", stdout);
	    else
		printf ("%s has ", user);

	    printf ("%d message%s (%d bytes)",
		    nmsgs, PLURALS(nmsgs), nbytes);
	}
	else
	    notifysw = 0;

	status = false;
    } else {
	if (notifysw & NT_NMAI)
	    printf (personal ? "You don't %s%s" : "%s doesn't %s",
		    personal ? "" : user, "have any mail waiting");
	else
	    notifysw = 0;
	status = true;
    }
    if (notifysw)
	printf (" on %s\n", host);

    return status;
}
